File: fe2-android/app/src/test/java/com/github/dedis/popstellar/repository/remote/GlobalNetworkManagerTest.java
package com.github.dedis.popstellar.repository.remote;

import com.github.dedis.popstellar.model.objects.Channel;
import com.github.dedis.popstellar.repository.LAORepository;
import com.github.dedis.popstellar.utility.handler.MessageHandler;
import com.github.dedis.popstellar.utility.scheduler.TestSchedulerProvider;
import com.google.gson.Gson;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.concurrent.TimeUnit;

import io.reactivex.Completable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.TestScheduler;
import io.reactivex.subjects.BehaviorSubject;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.*;

@RunWith(MockitoJUnitRunner.class)
public class GlobalNetworkManagerTest {

  @Mock LAORepository repository;
  @Mock MessageHandler handler;
  @Mock Gson gson;

  @Test
  public void initializationProducesAValidConnection() {
    TestSchedulerProvider schedulerProvider = new TestSchedulerProvider();
    TestScheduler testScheduler = schedulerProvider.getTestScheduler();
    ConnectionFactory factory = mock(ConnectionFactory.class);

    Connection firstConnection = mock(Connection.class);
    when(firstConnection.observeMessage()).thenReturn(BehaviorSubject.create());
    when(firstConnection.observeConnectionEvents()).thenReturn(BehaviorSubject.create());

    when(factory.createConnection(anyString())).thenReturn(firstConnection);

    GlobalNetworkManager networkManager =
        new GlobalNetworkManager(repository, handler, factory, gson, schedulerProvider);
    verify(factory).createConnection(anyString());

    Completable sendMessage = networkManager.getMessageSender().unsubscribe(Channel.ROOT);
    verify(firstConnection, never()).sendMessage(any());

    Disposable disposable = sendMessage.subscribe();
    testScheduler.advanceTimeBy(5, TimeUnit.SECONDS);
    disposable.dispose();

    verify(firstConnection).sendMessage(any());
  }

  @Test
  public void connectingToANewConnectionClosesTheLast() {
    ConnectionFactory factory = mock(ConnectionFactory.class);

    Connection firstConnection = mock(Connection.class);
    when(firstConnection.observeMessage()).thenReturn(BehaviorSubject.create());
    when(firstConnection.observeConnectionEvents()).thenReturn(BehaviorSubject.create());

    when(factory.createConnection(anyString())).thenReturn(firstConnection);

    GlobalNetworkManager networkManager =
        new GlobalNetworkManager(repository, handler, factory, gson, new TestSchedulerProvider());
    verify(factory).createConnection(anyString());

    Connection secondConnection = mock(Connection.class);
    when(secondConnection.observeMessage()).thenReturn(BehaviorSubject.create());
    when(secondConnection.observeConnectionEvents()).thenReturn(BehaviorSubject.create());

    when(factory.createConnection(anyString())).thenReturn(secondConnection);
    networkManager.connect("new url");

    verify(firstConnection).close();
  }
}

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/remote/GlobalNetworkManagerTest.kt
package com.github.dedis.popstellar.repository.remote

import com.github.dedis.popstellar.model.objects.Channel
import com.github.dedis.popstellar.testutils.MockitoKotlinHelpers
import com.github.dedis.popstellar.utility.handler.MessageHandler
import com.github.dedis.popstellar.utility.scheduler.TestSchedulerProvider
import com.google.gson.Gson
import io.reactivex.subjects.BehaviorSubject
import java.util.concurrent.TimeUnit
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.MockitoJUnitRunner

@RunWith(MockitoJUnitRunner::class)
class GlobalNetworkManagerTest {
  @Mock lateinit var handler: MessageHandler

  @Mock lateinit var gson: Gson

  @Test
  fun initializationProducesAValidConnection() {
    val schedulerProvider = TestSchedulerProvider()
    val testScheduler = schedulerProvider.testScheduler
    val factory = Mockito.mock(ConnectionFactory::class.java)
    val firstConnection = Mockito.mock(MultiConnection::class.java)

    Mockito.`when`(firstConnection.observeMessage()).thenReturn(BehaviorSubject.create())
    Mockito.`when`(firstConnection.observeConnectionEvents()).thenReturn(BehaviorSubject.create())
    Mockito.`when`(factory.createMultiConnection(ArgumentMatchers.anyString()))
      .thenReturn(firstConnection)

    val networkManager = GlobalNetworkManager(handler, factory, gson, schedulerProvider)
    Mockito.verify(factory).createMultiConnection(ArgumentMatchers.anyString())

    val sendMessage = networkManager.messageSender.unsubscribe(Channel.ROOT)
    Mockito.verify(firstConnection, Mockito.never()).sendMessage(MockitoKotlinHelpers.any())

    val disposable = sendMessage.subscribe()
    testScheduler.advanceTimeBy(5, TimeUnit.SECONDS)
    disposable.dispose()

    Mockito.verify(firstConnection).sendMessage(MockitoKotlinHelpers.any())
  }

  @Test
  fun connectingToANewConnectionClosesTheLast() {
    val factory = Mockito.mock(ConnectionFactory::class.java)
    val firstConnection = Mockito.mock(MultiConnection::class.java)

    Mockito.`when`(firstConnection.observeMessage()).thenReturn(BehaviorSubject.create())
    Mockito.`when`(firstConnection.observeConnectionEvents()).thenReturn(BehaviorSubject.create())
    Mockito.`when`(factory.createMultiConnection(ArgumentMatchers.anyString()))
      .thenReturn(firstConnection)

    val networkManager = GlobalNetworkManager(handler, factory, gson, TestSchedulerProvider())
    Mockito.verify(factory).createMultiConnection(ArgumentMatchers.anyString())

    val secondConnection = Mockito.mock(MultiConnection::class.java)
    Mockito.`when`(secondConnection.observeMessage()).thenReturn(BehaviorSubject.create())
    Mockito.`when`(secondConnection.observeConnectionEvents()).thenReturn(BehaviorSubject.create())
    Mockito.`when`(factory.createMultiConnection(ArgumentMatchers.anyString()))
      .thenReturn(secondConnection)

    networkManager.connect("new url")

    Mockito.verify(firstConnection).close()
  }
}
